### ISSUE TITLE
set up CI server  (like Travis) for testing

### ISSUE THREAD

**Issue by mikofski** [NREL/rdtools#76]
# Current State

* there is no continuous integration to test, build, and deploy rdtools
* there is an undue burden on maintainers, contributors, and users to manually perform tests on all platforms and Python versions that is error prone
* merging pull requests requires manual testing on different systems that may have unknown environments, package versions, `PATH` clashes, and other detritus that make testing flawed, producing false results
* there is an undue burden on maintainers to manually release new versions to PyPI, Anaconda, and GitHub releases

# Proposal

* set up a free [TravisCI](https://travis-ci.org/) account to test both Python 2.7 and Python 3.6
* since rdtools is pure python this server is sufficient for all platforms
* add a section to `setup.cfg` that sets wheels to be universal (py2 & py3)
    ```
    [bdist_wheel]
    universal = 1
    ```
* add a section to the `.travis.yml` file to build wheels and archives and then deploy them to GitHub and PyPI

# Future work

* add a `requirements.txt` file at the top level so that CI servers and IDE apps can infer the virtual 
environment or start using [`pipenv`](https://docs.pipenv.org/)
* set up conda forge to build and release rdtools to Anaconda automatically from GitHub

# examples of `.travis.yml`

* [pvmismatch](https://github.com/SunPower/PVMismatch/blob/master/.travis.yml)
* [pecos](https://github.com/sandialabs/pecos/blob/master/.travis.yml)
* [pvlib-python](https://github.com/pvlib/pvlib-python/blob/master/.travis.yml)
* [SolarUtils](https://github.com/SunPower/SolarUtils/blob/master/.travis.yml)
* [Carousel](https://github.com/SunPower/Carousel/blob/master/.travis.yml)
* [UncertaintyWrapper](https://github.com/SunPower/UncertaintyWrapper/blob/master/.travis.yml)

# Travis CI docs and CLI

* [docs](https://docs.travis-ci.com/)
* [CLI](https://github.com/travis-ci/travis.rb#installation) only necessary to create secure password hashes, alternately use the web UI to create hidden secure credentials

**Comment by abshinn** [comment]
CI is a great idea, @mikofski. Thanks for suggesting. I have experience with [Concourse](https://concourse-ci.org/) for private repositories. I agree that TravisCI is the best option here.

Do you have the bandwidth to set this up? If not, I am happy to take it on and loop you in for review.

I also want to point out that the [current release procedure](https://github.com/NREL/rdtools/wiki/Release-procedure) involves manually running both tests and a jupyter notebook. The notebook is painfully manual, but it serves two purposes: 1) it uses real-world data on the whole analysis pipeline, so it can be thought of as an integration test, and 2) ensures that the example notebook is up-to-date. It is great to provide an example notebook, but ideally the unit tests should provide sufficient coverage such that the integration test is not required for a release. I figure we should improve test coverage in parallel to getting CI set up.

**Comment by mdeceglie** [comment]
I went ahead and turned on Travis for [the repo](https://travis-ci.org/NREL/rdtools) and pushed a branch with some initial progress. 

# Workflow
We should settle on the implications of adding CI to our current workflow. I suggest that it may be preferable to continue to maintain the `development` branch separate from `master`. CI will help with integration into `development` and testing prior to mergint with `master`. But PRs to `master` would still be only those intended for a release (after testing of example notebooks by hand).

# Progress

Progress thus far is in the [`setup_CI` branch.](https://github.com/NREL/rdtools/tree/setup_CI) My intent was to try and keep it as simple as possible (for example, I don't think we need to get conda involved) but accomplish the same tests we were [running manually](https://github.com/NREL/rdtools/wiki/Release-procedure), in particular testing both minimum and up-to-date versions for the dependencies.

In it's current state the build for the `setup_CI` branch fails. This is because Travis is not setting up the environment variable `REQ_FILE`. I don't understand why not. I would expect the log to show `export REQ_FILE=...` as it does for example with `FLASK` in [Flask-Split](https://travis-ci.org/jpvanhal/flask-split/jobs/138204623) which has a pretty similar structure to its [`.travis.yml`](https://github.com/jpvanhal/flask-split/blob/master/.travis.yml) 

I'm also not thrilled with the manual synchronization that we would have to maintain between the requirements files in `./ci/` and `setup.py`.

@abshinn, you indicated he would be interested in looking into CI, so I'll let you take the lead from here.

**Comment by computron** [comment]
I think this is already fixed, but if you see more problems with Travis in the future, we've had pretty good experience with CircleCI in the past
